valohai_yaml/objs/workload_resources.py

```diff
@@ -0,0 +1,82 @@
+from collections import OrderedDict
+
+from valohai_yaml.objs.base import Item
+from valohai_yaml.types import SerializedDict
+
+
+class WorkloadResources(Item):
+    """
+    Represents a workload resource definition.
+
+    Resources include: cpu, memory, devices.
+    Both the resources and their properties are optional.
+    """
+
+    def __init__(
+        self,
+        *args: OrderedDict,
+    ) -> None:
+        self.cpu: ResourceCPU | None = None
+        self.memory: ResourceMemory | None = None
+        self.devices: ResourceDevices | None = None
+        self._parse_args(args[0])
+
+    def __repr__(self) -> str:
+        """Resources contents."""
+        return f'WorkflowResources("cpu": {self.cpu}, "memory": {self.memory}, "devices": {self.devices})'
+
+    def _parse_args(self, resources: dict) -> None:
+        if not resources:
+            return
+
+        if cpu := resources.get("cpu"):
+            self.cpu = ResourceCPU(cpu)
+        if memory := resources.get("memory"):
+            self.memory = ResourceMemory(memory)
+        if devices := resources.get("devices"):
+            self.devices = ResourceDevices(devices)
```

Comment:
on all of the `Item` subclasses we have defined here, the interface is a bit off

+ the constructor `__init__` should take arguments in their final form, not a dict blob that you parse, probably the right interface in here would be `cpu: ResourceCPU, memory: ResourceMemory, devices: ResourceDevices`
  + example: https://github.com/valohai/valohai-yaml/blob/master/valohai_yaml/objs/step.py
+ you shouldn't implement custom `_parse_args` but either use the superclass `parse` class method or override it with your own e.g.
  + base: https://github.com/valohai/valohai-yaml/blob/master/valohai_yaml/objs/base.py#L42-L52
  + override: https://github.com/valohai/valohai-yaml/blob/master/valohai_yaml/objs/step.py#L70-L80

you can take more inspiration from most of the other `Item` subclasses defined in the `objs` directory e.g.

+ https://github.com/valohai/valohai-yaml/blob/master/valohai_yaml/objs/endpoint.py
+ https://github.com/valohai/valohai-yaml/blob/master/valohai_yaml/objs/task.py
+ https://github.com/valohai/valohai-yaml/blob/master/valohai_yaml/objs/parameter.py
+ https://github.com/valohai/valohai-yaml/blob/master/valohai_yaml/objs/environment_variable.py (uses superclass `parse` as not defined)

Reply:
Right, on it.

Reply:
Items refactored.